File: dailyQuestion/2020-05/05-28/python/solution_recursion.py

ptr = 0

def decodeString(s: str) -> str :
    """

    :param s:
    :return:
    """
    def getDigits() :
        global ptr
        global s
        ret = 0
        while s[ptr] >= "0" and s[ptr] <= "9" :
            ret = ret * 10 + int(s[ptr])
            ptr += 1
        return ret

    def getString() :
        global s
        global ptr
        if len(s) == ptr or s[ptr] == "]" :
            return ""
        current = s[ptr]
        result = ""
        if current >= "0" and current <= "9" :
            repeat = getDigits()
            ptr += 1
            s1 = getString()
            ptr += 1
            result = s1 * repeat
        elif (current >= "a" and current <= "z") or current >= "A" and current <= "A" :
            result = current
            ptr += 1
        return result + getString()

    return getString()

if __name__ == "__main__" :
    s = "12[a]2[bc]"
    result = decodeString(s)
    print(result)

File: dailyQuestion/2020/2020-05/05-28/golang/solution_recursion.go
package main

import (
	"fmt"
	"strings"
)

func main() {
	s := "12[a]12[bc]"
	result := decodeString(s)
	fmt.Println(result)
}

var (
	src string
	ptr int
)

func decodeString(s string) string {
	src = s
	ptr = 0
	return getString()
}

func getString() string {
	if ptr == len(src) || src[ptr] == ']' {
		return ""
	}
	repeat := 1
	result := ""
	current := src[ptr]
	if current >= '0' && current <= '9' {
		repeat = getDigits()
		ptr++
		s := getString()
		ptr++
		result = strings.Repeat(s, repeat)
	} else if current >= 'a' && current <= 'z' || current >= 'A' && current <= 'Z' {
		result = string(current)
		ptr++
	}
	return result + getString()

}

func getDigits() int {
	ret := 0
	for ; src[ptr] >= '0' && src[ptr] <= '9'; ptr++ {
		ret = ret*10 + int(src[ptr]-'0')
	}
	return ret
}
